registry/registry.go

```diff
@@ -0,0 +1,359 @@
+	if err != nil {
+		return nil, fmt.Errorf("error configuring logger: %v", err)
+	}
+
+	// inject a logger into the uuid library. warns us if there is a problem
+	// with uuid generation under low entropy.
+	uuid.Loggerf = ctxu.GetLogger(ctx).Warnf
+
+	app := handlers.NewApp(ctx, config)
+	// TODO(aaronl): The global scope of the health checks means NewRegistry
+	// can only be called once per process.
+	app.RegisterHealthChecks()
+	handler := configureReporting(app)
+	handler = alive("/", handler)
+	handler = health.Handler(handler)
+	handler = panicHandler(handler)
+	handler = gorhandlers.CombinedLoggingHandler(os.Stdout, handler)
+
+	server := &http.Server{
+		Handler: handler,
+	}
+
+	ln, err := listener.NewListener(config.HTTP.Net, config.HTTP.Addr)
+	if err != nil {
+		return nil, err
+	}
+
+	var debugLn net.Listener
+	if config.HTTP.Debug.Addr != "" {
+		debugLn, err = listener.NewListener("tcp", config.HTTP.Debug.Addr)
+		if err != nil {
+			return nil, fmt.Errorf("error listening on debug interface: %v", err)
+		}
+		log.Infof("debug server listening %v", config.HTTP.Debug.Addr)
+	}
+
+	if config.HTTP.TLS.Certificate != "" {
+		tlsConf := &tls.Config{
+			ClientAuth:               tls.NoClientCert,
+			NextProtos:               []string{"http/1.1"},
+			Certificates:             make([]tls.Certificate, 1),
+			MinVersion:               tls.VersionTLS10,
+			PreferServerCipherSuites: true,
+			CipherSuites: []uint16{
+				tls.TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256,
+				tls.TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256,
+				tls.TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA,
+				tls.TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA,
+				tls.TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA,
+				tls.TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA,
+				tls.TLS_RSA_WITH_AES_128_CBC_SHA,
+				tls.TLS_RSA_WITH_AES_256_CBC_SHA,
+			},
+		}
+
+		tlsConf.Certificates[0], err = tls.LoadX509KeyPair(config.HTTP.TLS.Certificate, config.HTTP.TLS.Key)
+		if err != nil {
+			return nil, err
+		}
+
+		if len(config.HTTP.TLS.ClientCAs) != 0 {
+			pool := x509.NewCertPool()
+
+			for _, ca := range config.HTTP.TLS.ClientCAs {
+				caPem, err := ioutil.ReadFile(ca)
+				if err != nil {
+					return nil, err
+				}
+
+				if ok := pool.AppendCertsFromPEM(caPem); !ok {
+					return nil, fmt.Errorf("Could not add CA to pool")
+				}
+			}
+
+			for _, subj := range pool.Subjects() {
+				ctxu.GetLogger(app).Debugf("CA Subject: %s", string(subj))
+			}
+
+			tlsConf.ClientAuth = tls.RequireAndVerifyClientCert
+			tlsConf.ClientCAs = pool
+		}
+
+		ln = tls.NewListener(ln, tlsConf)
+		ctxu.GetLogger(app).Infof("listening on %v, tls", ln.Addr())
+	} else {
+		ctxu.GetLogger(app).Infof("listening on %v", ln.Addr())
+	}
+
+	return &Registry{
+		app:     app,
+		config:  config,
+		server:  server,
+		ln:      ln,
+		debugLn: debugLn,
+	}, nil
+}
+
+// Serve runs the registry's HTTP server(s).
+func (registry *Registry) Serve() error {
+	defer registry.ln.Close()
+
+	errChan := make(chan error)
```

Comment:
True, though the way this is used by CmdMain, the first error will cause the program to exit. I'll change this to a buffered channel to avoid the goroutine leak in case other programs embed this in a way where they don't exit after an error.